### ISSUE TITLE
Special View selection vs. inheritance

### ISSUE THREAD

**Issue by nomeata** [clientIO/joint#168]
Currently, the code that selects the right view for a cell is this:
```
    createViewForModel: function(cell) {

        var view;

        var type = cell.get('type');
        var module = type.split('.')[0];
        var entity = type.split('.')[1];

        // If there is a special view defined for this model, use that one instead of the default `elementView`/`linkView`.
        if (joint.shapes[module] && joint.shapes[module][entity + 'View']) {

            view = new joint.shapes[module][entity + 'View']({ model: cell, interactive: this.options.interactive });

        } else if (cell instanceof joint.dia.Element) {

            view = new this.options.elementView({ model: cell, interactive: this.options.interactive });

        } else {

            view = new this.options.linkView({ model: cell, interactive: this.options.interactive });
        }

        return view;
    },
```

So if I want a certain element to be rendered by a special view, I need to put the view in `joint.shapes.foo.barView` and set `type = "foo.bar"`.

But if I now derive from this element class, and change the type (as it seems to be best practice) to, say, `foo.baz`, then it will not use the special view any more.

It seems that it should not just take the `type` of the cell at hand, but also go through its parents in the inheritance tree and see if there is a special view for them – if that is possible, given that `type` is set in defaults.

**Comment by nomeata** [comment]
One solution seems to be to add `joint.shapes.foo.bazView = joint.shapes.foo.barView;`, i.e. replicate the model object inheritance tree on the view side.

**Comment by kumilingus** [comment]
Fixed in 62b444f0b1473331b7758a95c83c8c05a35d5722.

It's possible now to pass a function to `elementView`/`linkView` paper option and decide what view to use based on the model.

```
var paper = new joint.dia.Paper({
  elementView: function(cell) {
    if (cell instanceof joint.shapes.foo.Bar) {
      return joint.shapes.foo.barView
    }
    return joint.dia.Element;
  }
})
```

It also introduce custom namespaces for cells and cellViews (doesn't have to be `joint.shapes` necessary).
```
    var paper = new joint.dia.Paper({
      cellNamespace: {
         foo: { barView: FooView, bazView: BazView }
      }
    })

   var graph = new joint.dia.Graph({ graphName: 'MyGraph' }, { cellNamespace: {
       foo: { bar: FooModel, baz: BazModel }
   }});
```

The cell type is basically a path with '.' delimeter. e.g.

    { type: 'foo.bar.baz' } // ==>
    namespace = `{ foo: { bar: { bazView: [will use this view] }}}`

Please check the commit for more informations. There is no documentation yet.

Thank you for your contribution.
